### PR TITLE
FIX: check if axes is off page before repositioning title

### LINKED ISSUE #18283: Automatic title placement wrong if parent axes is off the page
### Bug report

**Bug summary**
When calculating the extent of a subplot axes, its contents depend on whether  title has been set in the axes below, i. e. the unrelated title bleeds into the extent and gets exported when saving while specifying the `bbox_inches`.

The extent itself doesn't seem to be the problem, but the inclusion of artefacts that should not be included in the plot, and aren't visible in the full figure (in the bounding box) are included and saved when giving an explicit bounding box to `fig.savefig`. 

**Code for reproduction**

```python
import matplotlib.pyplot as plt
fig, axs = plt.subplots(2, 1)
axs[0].plot([1,2], [3,4])
extent = axs[0].get_window_extent().transformed(fig.dpi_scale_trans.inverted())
fig.savefig('/tmp/expected.png', bbox_inches=extent)
axs[1].set_title('pls ignore me')
# extents are actually unchanged
extent2 = axs[0].get_window_extent().transformed(fig.dpi_scale_trans.inverted())
fig.savefig('/tmp/actual.png', bbox_inches=extent2)
fig.savefig('/tmp/full.png')
```

**Actual outcome**
The title of `axs[1]` is shown.

actual

**Expected outcome**
 It should not be included, as visible in the full picture.

expected
full

**Matplotlib version**

  * Operating system:
  * Matplotlib version: 3.2.1
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.8.2

Environment has been installed with the default conda channel.

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #18283

We now automatically place the title to avoid decorators on the x axis (particuarly if the axis is at the top).  However if the whole axes is cropped due to a `bbox_inches=extent` call, then the title still gets placed at a default `y=0`.  This change checks if the `top` is in the figure, and if not, then it doesn't auto position the title. 

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
yeeesh, why am I getting flake8 complaints for a file I didn't edit?

**Comment by tacaswell** [comment]
It looks like something changed in the flake8 incantation?

File: lib/matplotlib/tests/test_pickle.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 from io import BytesIO
 import pickle
-import platform
```
Comment (jklymak) [inline lib/matplotlib/tests/test_pickle.py:3]: This is just to make flake8 happy - not part of this PR, but flake8 now seems to check all files, not just modified.

File: lib/matplotlib/tests/test_axes.py [hunk 2]
```diff
@@ -5533,6 +5533,19 @@ def test_title_xticks_top_both():
     assert ax.title.get_position()[1] > 1.04
 
 
+def test_title_no_move_off_page():
+    # If an axes is off the figure (ie. if its cropped during a save)
```
Comment (dopplershift) [inline lib/matplotlib/tests/test_axes.py:5537]: "its" -> "it's"

**Review by timhoffm — APPROVED** [review]
